src/Common/HostResolvePool.cpp

```diff
@@ -113,6 +123,7 @@ void HostResolver::updateWeights()
 
     if (getTotalWeight() == 0 && !records.empty())
     {
+        CurrentMetrics::sub(metrics.banned_count, records.size());
```

Comment:
Some of records may be alive here, need to decrease only for records with failed==true, I guess.
Like
```
size_t failed_total = 0;
for (auto & rec : record)
{
  if (rec.failed)
  {
    ++failed_total;
    rec.failed = false;
  }
}
CurrentMetrics::sub(metrics.banned_count, failed_total);
```

Reply:
We are here only when we need to reset failed flag for all records.

Reply:
`updateWeights` is called in `updateImpl` too.

Reply:
Sorry I do not understand what you are pointing at.

`update` fetch new hosts from resolver cache and calls `updateImpl` under a lock.
`updateImpl` applys new set of IP's to the inner stage and calls `updateWeights`.
`updateWeights` firstly calls `updateWeightsImpl`. `updateWeightsImpl` calculate new weights for random choises.
after that `updateWeights` checks if we seceded in that weight calculation. If all hosts are benned as faulty then we reset faulty flags for all the records. It is our only choice here. And after `updateWeights` calls `updateWeightsImpl` one more time.

Reply:
`updateWeights` is called in two cases -- 1. we do `update`. 2. Some records changed their weight.

Reply:
Ah, I missed that getTotalWeight() == 0 when this code call.
Sorry,my mistake.